src/dbus_api/pool.rs

```diff
@@ -360,48 +360,75 @@ fn get_pool_name(i: &mut IterAppend, p: &PropInfo<MTFn<TData>, TData>) -> Result
 }
 
-fn get_pool_total_physical_used(
-    i: &mut IterAppend,
-    p: &PropInfo<MTFn<TData>, TData>,
-) -> Result<(), MethodErr> {
-    fn get_used((_, uuid, pool): (Name, PoolUuid, &dyn Pool)) -> Result<String, MethodErr> {
-        let err_func = |_| {
-            MethodErr::failed(&format!(
-                "no total physical size computed for pool with uuid {}",
-                uuid
-            ))
-        };
-
-        pool.total_physical_used()
-            .map(|u| Ok(format!("{}", *u)))
-            .map_err(err_func)?
-    }
+fn get_all_properties(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
+    let message: &Message = m.msg;
+    let object_path = &m.path;
+    let dbus_context = m.tree.get_data();
+    let engine = dbus_context.engine.borrow();
 
-    get_pool_property(i, p, get_used)
-}
+    let default_return: HashMap<String, (bool, Variant<Box<dyn RefArg>>)> = HashMap::new();
+    let return_message = message.method_return();
 
-fn get_pool_total_physical_size(
-    i: &mut IterAppend,
-    p: &PropInfo<MTFn<TData>, TData>,
-) -> Result<(), MethodErr> {
-    get_pool_property(i, p, |(_, _, p)| {
-        Ok(format!("{}", *p.total_physical_size()))
-    })
-}
+    let uuid = get_data!(object_path; default_return; return_message).uuid;
+    let (_, pool) = get_pool!(engine; uuid; default_return; return_message);
 
-fn get_pool_state(i: &mut IterAppend, p: &PropInfo<MTFn<TData>, TData>) -> Result<(), MethodErr> {
-    get_pool_property(i, p, |(_, _, pool)| Ok(pool.state() as u16))
-}
+    let mut return_value: HashMap<String, (bool, Variant<Box<dyn RefArg>>)> = HashMap::new();
+    return_value.insert(
+        consts::POOL_TOTAL_SIZE_PROP.to_string(),
+        (
+            true,
+            Variant(Box::new(pool.total_physical_size().to_string())),
+        ),
+    );
 
-fn get_pool_extend_state(
-    i: &mut IterAppend,
-    p: &PropInfo<MTFn<TData>, TData>,
-) -> Result<(), MethodErr> {
-    get_pool_property(i, p, |(_, _, pool)| Ok(pool.extend_state() as u16))
+    Ok(vec![return_message.append3(
+        return_value,
+        msg_code_ok(),
+        msg_string_ok(),
+    )])
 }
 
-fn get_space_state(i: &mut IterAppend, p: &PropInfo<MTFn<TData>, TData>) -> Result<(), MethodErr> {
-    get_pool_property(i, p, |(_, _, pool)| Ok(pool.free_space_state() as u16))
+fn get_properties(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
+    let message: &Message = m.msg;
+    let mut iter = message.iter_init();
+    let object_path = &m.path;
+    let dbus_context = m.tree.get_data();
+    let engine = dbus_context.engine.borrow();
+    let properties: Array<String, _> = get_next_arg(&mut iter, 0)?;
+
+    let default_return: HashMap<String, PropertyReturn> = HashMap::new();
+    let return_message = message.method_return();
+
+    let uuid = get_data!(object_path; default_return; return_message).uuid;
+    let (_, pool) = get_pool!(engine; uuid; default_return; return_message);
+
```

Comment:
Please turn the array of properties into a set before filling the map. My goal is just to prevent some frivolously obnoxious person from requesting the same expensive value to the limit of the size of the D-Bus array. If there's a way when filling the HashMap to do this as well, so that a request for a value can be guaranteed to be satisfied exactly once, that's great to.

Reply:
I'm a little confused about the benefit of this. This still doesn't stop the user from supplying a bunch of random values which could still cause problems. Is the rationale that this would at least prevent the engine from running the code to fetch actual properties multiple times?

Reply:
Thinking further, I don't think it's necessary to return any entry for an unrecognized key. Let the fact that there is no entry for the key in the table be the indicator that the key was unrecognized.

Then the client does:
```
result = map.get(key)
if result is None:
   # raise some exception, we have a bug OR ...
(value_exists, value) = result
if !value_exists:
   # do reasonable action
else:
  # got the value, so use it
```

That makes the type of the table simpler, guarantees that it is no larger than the size of the set of keys that the daemon does recognize, and makes the type of the table the same as that for ```get_all_properties```.

Regarding your previous question, 2^20 keys will require only the computation necessary to realize that they are unrecognized, and in my new formulation won't lead to a very large table, which is also good. But, if the same recognized key is requested 2^20 times, I'ld like to avoid computing the value for that key 2^20 times, and just compute it once.

Reply:
Resolved using `itertools`.